"Exit handler never called" error when /proc is not accessible

### Is there an existing issue for this?

- [X] I have searched the existing issues

### This issue exists in the latest npm version

- [X] I am using the latest npm

### Current Behavior

When `/proc/stat` and `/proc/cpuinfo` are not accessible and version 1 `package-lock.json` is present, `npm ci` fails to install packages with "exit handler never called" error after "The package-lock.json file was created with an old version of npm" warning. This is caused by a bug in promise-call-limit library.

To check whether the system is affected by the bug one can run this command: 

```bash
node -e "console.log(require('os').cpus())"
```

If it prints an empty list `[]` then there is a bug and npm will fail either at `idealTree:inflate` stage or at `build:run:*` stage. This is caused by a bug in `callLimit` function that doesn't run any of the passed functions if CPU count cannot be determined. This function is invoked in two places: [when parsing old lock file](https://github.com/npm/cli/blob/54795a3554691e97b369baac714cb43f4342af8a/workspaces/arborist/lib/arborist/build-ideal-tree.js#L748) and [when running build:run](https://github.com/npm/cli/blob/54795a3554691e97b369baac714cb43f4342af8a/workspaces/arborist/lib/arborist/rebuild.js#L322).

I have made a pull request to fix the bug: https://github.com/isaacs/promise-call-limit/pull/11 . Please consider updating promise-call-limit version.

So, the bug only happens if /proc not being accessible and stage is `idealTree:inflate` (and old lockfile is present) or `build:run:*`.

Searching around shows that there are bugs from users trying to use npm in different OSes which might be caused by the same reason: 

- https://github.com/npm/cli/issues/5488 (bug on idealTree:inflate stage when parsing lockfile version 1 on Android)
- https://github.com/npm/cli/issues/4839 (bug on build:run:postinstall stage in OpenBSD)
- https://github.com/npm/cli/issues/4719 (bug when parsing lockfile version 1 but OS is not specified by user)
- https://github.com/npm/cli/issues/6213 (old lockfile, OS unknown)
- https://github.com/npm/cli/issues/6076 (old lockfile, OS unknown)
- https://github.com/npm/cli/issues/6017 (old lockfile, Android)
- https://github.com/npm/cli/issues/5886 (Android, log not published)
- https://github.com/npm/cli/issues/5841 (Android, build:run:install stage)
- https://github.com/npm/cli/issues/5080 (Android and old lock file)
- there are more closed issues but I'm a bit tired of looking through them

These users should run the command above to check whether their issue is caused by this bug.



### Expected Behavior

npm ci runs successfully without errors

### Steps To Reproduce

1. Create a container without /proc mounted or unmount /proc or use OS without /proc
2. Run `node -e "console.log(require('os').cpus())"` and see that it prints `[]`
3. Create a directory and create a package-lock.json with following contents:
```json
{
  "name": "test-npm-bug",
  "requires": true,
  "lockfileVersion": 1,
  "dependencies": {
    "@nodelib/fs.stat": {
      "version": "2.0.5",
      "resolved": "https://registry.npmjs.org/@nodelib/fs.stat/-/fs.stat-2.0.5.tgz",
      "integrity": "sha512-RkhPPp2zrqDAQA/2jNhnztcPAlv64XdhIp7a7454A5ovI7Bukxgt7MX7udwAu3zg1DcpPU0rz3VV1SeaqvY4+A=="
    }
  }
}

```

Also create package.json: 

```json
{
  "name": "test-npm-bug",
  "dependencies": {
      "@nodelib/fs.stat": "2.0.5"
  }
}
```

4. Run npm ci --ddd
5. See messages in log: 

```
npm WARN old lockfile 
npm WARN old lockfile The package-lock.json file was created with an old version of npm,
npm WARN old lockfile so supplemental metadata must be fetched from the registry.
npm WARN old lockfile 
npm WARN old lockfile This is a one-time fix-up, please be patient...
npm WARN old lockfile 
npm verb unfinished npm timer command:ci 1680276840213
npm verb unfinished npm timer idealTree 1680276840284
npm verb unfinished npm timer idealTree:inflate 1680276840288
npm info ok 
npm ERR! Exit handler never called!
```


### Environment

- npm: 9.6.3
- Node.js: v18.15.0
- OS Name: Linux
- System Model Name: 
- npm config:
```ini
; node bin location = /usr/bin/node
; node version = v18.15.0
; npm local prefix = /shared/test-npm
; npm version = 9.6.3
; cwd = /shared/test-npm
; HOME = (removed)
; Run `npm config ls -l` to show all defaults.
```

--- a/package.json
+++ b/package.json
@@ -68,7 +68,7 @@
     "columnify": "^1.6.0",
     "fastest-levenshtein": "^1.0.16",
     "fs-minipass": "^3.0.1",
-    "glob": "^9.3.1",
+    "glob": "^9.3.2",
     "graceful-fs": "^4.2.11",
     "hosted-git-info": "^6.1.1",
     "ini": "^3.0.1",

--- a/workspaces/arborist/package.json
+++ b/workspaces/arborist/package.json
@@ -30,7 +30,7 @@
     "parse-conflict-json": "^3.0.0",
     "proc-log": "^3.0.0",
     "promise-all-reject-late": "^1.0.0",
-    "promise-call-limit": "^1.0.1",
+    "promise-call-limit": "^1.0.2",
     "read-package-json-fast": "^3.0.2",
     "semver": "^7.3.7",
     "ssri": "^10.0.1",

